### PR TITLE
Switch idling over to using owner references.

### LINKED ISSUE #17798: Update the idler to use owner refs
The idler may be broken today.

### PR DESCRIPTION
**Pull request by DirectXMan12**
This commit switches the `oc idle` code over to using owner references
instead of the created-by annotation and the OpenShift-specific
deployment config annotations.

Closes #17798

### CODE REVIEW

**Comment by sjenning** [comment]
/assign @soltysh

xref https://bugzilla.redhat.com/show_bug.cgi?id=1529473

Thanks for the fix.  Not familiar with this area of the code, but did a sanity review and looks fine to me.  I'll lgtm when tests are green.

**Comment by DirectXMan12** [comment]
doing some sanity tests, I think I may have missed one thing, should have a new version up shortly

**Comment by DirectXMan12** [comment]
hmm, the thing I found is just a bug in the way we handle kube deployements...

**Comment by DirectXMan12** [comment]
I'll follow up with another PR for that so we can get this fix merged

/retest

**Comment by DirectXMan12** [comment]
Turns out passing an `core/__internal__` RESTClient to an `extensions/v1beta1` client constructor now results in a Frankensteinian monster of a client with a tendency to produce incorrect URLs.  Let's not do that any more.

(eventually, I'll swap out the proper scale client, and it won't matter, but for the mean time, we need it)

File: pkg/oc/cli/cmd/idle.go [hunk 1]
```diff
@@ -294,37 +291,29 @@ func (o *IdleOptions) calculateIdlableAnnotationsByService(f *clientcmd.Factory)
...
-	annotations := objMeta.GetAnnotations()
+	ownerRefs := objMeta.GetOwnerReferences()
 
-	creatorRefRaw, creatorListed := annotations[api.DeprecatedKubeCreatedByAnnotation]
-	if !creatorListed {
-		// if we don't have a creator listed, try the openshift-specific Deployment annotation
-		dcName, dcNameListed := annotations[appsapi.DeploymentConfigAnnotation]
-		if !dcNameListed {
-			return nil, nil
+	for _, ownerRef := range ownerRefs {
```
Comment (soltysh) [inline pkg/oc/cli/cmd/idle.go]: Replace with `GetControllerOf` from `vendor/k8s.io/apimachinery/pkg/apis/meta/v1/controller_ref.go`

File: pkg/oc/cli/cmd/idle_test.go [hunk 2]
```diff
@@ -4,54 +4,39 @@ import (
...
-	}
-
+	vTrue := true
 	return kapi.Pod{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      name,
 			Namespace: "somens",
-			Annotations: map[string]string{
-				oapi.DeprecatedKubeCreatedByAnnotation: string(createdByRefJson),
+			OwnerReferences: []metav1.OwnerReference{
```
Comment (soltysh) [inline pkg/oc/cli/cmd/idle_test.go]: Replace with `NewControllerRef` from `vendor/k8s.io/apimachinery/pkg/apis/meta/v1/controller_ref.go`

File: pkg/oc/cli/cmd/idle_test.go [hunk 3]
```diff
@@ -60,25 +45,14 @@ func makeRC(name, dcName, createdByDCName string, t *testing.T) *kapi.Replicatio
...
 
-	if createdByDCName != "" {
-		codec := legacyscheme.Codecs.LegacyCodec(schema.GroupVersion{Group: kapi.GroupName, Version: "v1"})
-		createdByRefJson, err := kruntime.Encode(codec, &kapi.SerializedReference{
-			Reference: kapi.ObjectReference{
-				Kind:      "DeploymentConfig",
-				Name:      createdByDCName,
-				Namespace: "somens",
+	if dcName != "" {
+		rc.OwnerReferences = []metav1.OwnerReference{
```
Comment (soltysh) [inline pkg/oc/cli/cmd/idle_test.go]: Ditto.

**Comment by soltysh** [comment]
> Turns out passing an core/__internal__ RESTClient to an extensions/v1beta1 client constructor now results in a Frankensteinian monster of a client with a tendency to produce incorrect URLs. Let's not do that any more.

Right, you need to use specific client for the constructor 😉

File: pkg/oc/cli/cmd/idle.go (hunk 1 above)
Comment (DirectXMan12) [reply pkg/oc/cli/cmd/idle.go]: ack

**Comment by sjenning** [comment]
flake https://github.com/openshift/origin/issues/17901
/retest

**Comment by sjenning** [comment]
@soltysh tests are green.  Can we get approval if all review comments have been addressed?

File: pkg/oc/cli/cmd/idle.go [hunk 4]
```diff
@@ -363,33 +331,26 @@ func findScalableResourcesForEndpoints(endpoints *kapi.Endpoints, decoder runtim
...
 		}
 
 		if controller != nil {
-			var parentControllerRef *kapi.ObjectReference
-			parentControllerRef, err = getControllerRef(controller, decoder)
-			if err != nil {
-				return nil, fmt.Errorf("unable to load the creator of %s %q: %v", controllerRef.Kind, controllerRef.Name, err)
-			}
-
+			var parentControllerRef *metav1.OwnerReference
```
Comment (soltysh) [inline pkg/oc/cli/cmd/idle.go]: Next time `parentControllerRef := metav1.GetControllerOf(controller)` you'll save yourself a few precious key strokes 😉

Comment (DirectXMan12) [reply pkg/oc/cli/cmd/idle.go]: lol, yeah, I think that was left over from before I refactored it a bit, and I needed the pre-assignment declaration, and never simplified after

**Pull request merged**